pkg/controllers/counter/controller.go

```diff
@@ -0,0 +1,127 @@
+/*
+Licensed under the Apache License, Version 2.0 (the "License");
+you may not use this file except in compliance with the License.
+You may obtain a copy of the License at
+
+    http://www.apache.org/licenses/LICENSE-2.0
+
+Unless required by applicable law or agreed to in writing, software
+distributed under the License is distributed on an "AS IS" BASIS,
+WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+See the License for the specific language governing permissions and
+limitations under the License.
+*/
+package counter
+
+import (
+	"context"
+	"fmt"
+	"time"
+
+	v1 "k8s.io/api/core/v1"
+	"k8s.io/apimachinery/pkg/api/errors"
+	"k8s.io/apimachinery/pkg/api/resource"
+	"k8s.io/apimachinery/pkg/types"
+	"knative.dev/pkg/logging"
+	controllerruntime "sigs.k8s.io/controller-runtime"
+
+	"sigs.k8s.io/controller-runtime/pkg/client"
+	"sigs.k8s.io/controller-runtime/pkg/controller"
+	"sigs.k8s.io/controller-runtime/pkg/event"
+	"sigs.k8s.io/controller-runtime/pkg/handler"
+	"sigs.k8s.io/controller-runtime/pkg/manager"
+	"sigs.k8s.io/controller-runtime/pkg/predicate"
+	"sigs.k8s.io/controller-runtime/pkg/reconcile"
+	"sigs.k8s.io/controller-runtime/pkg/source"
+
+	"github.com/awslabs/karpenter/pkg/apis/provisioning/v1alpha5"
+)
+
+// Controller for the resource
+type Controller struct {
+	ctx        context.Context
+	kubeClient client.Client
+}
+
+// NewController is a constructor
+func NewController(ctx context.Context, kubeClient client.Client) *Controller {
+	return &Controller{
+		ctx:        ctx,
+		kubeClient: kubeClient,
+	}
+}
+
+// Reconcile a control loop for the resource
+func (c *Controller) Reconcile(ctx context.Context, req reconcile.Request) (reconcile.Result, error) {
+	logging.FromContext(ctx).Infof("Updating resource counts on provisioner")
+	// Retrieve the provisioner
+	provisioner := &v1alpha5.Provisioner{}
+	if err := c.kubeClient.Get(ctx, req.NamespacedName, provisioner); err != nil {
+		if !errors.IsNotFound(err) {
+			return reconcile.Result{}, err
+		}
+		return reconcile.Result{}, nil
+	}
+	persisted := provisioner.DeepCopy()
+
+	// Determine resource usage and update provisioner.status.resources
+	resourceCounts, err := c.resourceCountsFor(ctx, provisioner.Name)
+	if err != nil {
+		return reconcile.Result{}, err
+	}
+
+	provisioner.Status.Resources = resourceCounts
+	if err := c.kubeClient.Status().Patch(ctx, provisioner, client.MergeFrom(persisted)); err != nil {
+		return reconcile.Result{}, fmt.Errorf("failed to persist changes to %s, %w", req.NamespacedName, err)
+	}
+
+	// Refresh the reconciler state values every 5 minutes irrespective of node events
+	return reconcile.Result{RequeueAfter: 5 * time.Minute}, nil
```

Comment:
Do we need this? Watch should be enough.

Reply:
Fair enough, I'll remove